### ISSUE TITLE
Allow expanding of the &-something in selector definition

### ISSUE THREAD

**Issue by sgtpep** [curvedmark/roole#38]
source:

```
.foo
  &-bar
    prop: value
```

output:

```
.foo-bar
  prop: value
```

Maybe it's harmful for some reason, because it's doesn't work so in SASS, but works in LESS, Stylus. But I personally got used to it and hope it'll be implemented in roole.

**Comment by curvedmark** [comment]
Interesting. Could you provide some real-world use case? If I'm not wrong, are you trying to achieve some kind of namespace for the id/class selectors (e.g., prepend something like `foo-` before every id/class selector)?

**Comment by iAdramelk** [comment]
+1 My real life example: https://github.com/jetstyle/jam-boilerplate/blob/master/blocks/b-text/b-text.styl (can't show more, they are in private repos but there is a hundreds of them).

We use it in every project in every file. (We use BEM methodology to name classes and use it to not repeat names, instead of writing `.b-block__element` we write `&__element` ). 

Read about BEM here:

http://coding.smashingmagazine.com/2012/04/16/a-new-front-end-methodology-bem/
http://bem.info/method/

**Comment by sgtpep** [comment]
Yes, I meant namespacing or some kind of prefixing. They are widely used in some CSS methodologies, such as [BEM](http://bem.info/) or [SMACSS](http://smacss.com/). It's great for scoping and isolating selectors while their number grows, so you lower the chances to accidentally override some existing selectors.

**Comment by curvedmark** [comment]
I'm aware of these two methodologies and had been designing a syntax to support selector namespacing/modulation, but didn't keep it in the initial release as it didn't work quite well.

Allowing `&` in the middle of a subselector seems to be a way to do that, but it has a few drawbacks:
- making the language more obscure
  
  ```
  .submit
      &button
          border: none
  ```
  
  did the author mean `.submitbutton` or `.submit button` but failed to include a space combinator?
- if you decide to use it, most selectors in your files now start with `&`, which is pretty ugly IMHO.

Also, implementing it does likely to incur some performance penalty, unless class selectors are the only selectors allowed to include `&`, which will be very weird.

My original idea was to use constructs like:

```
@module foo
    .bar
        display: none
```

which will compile to

```
.foo-bar {
    display: none;
}
```

so you can write your selectors in the simplest form. But this syntax is not without problems (e.g., How do you specify a selector that don't need to be namespaced? How do you specify a selector whose value is the module name? Only class selectors will be namespaced? Why use `-` as a separator? etc).

If somehow the selectors can be namespaced elegantly, I'd be happy to implement it.

**Comment by sgtpep** [comment]
I'm afraid there are no any elegant solution, since CSS does not support namespacing natively, so developer is forced to stick with class names, dashes (someone uses underscores, but it's ugly) and some homebrew convention for writing them. I realize your point about obscuring the language, but I think it should be up to the developer to use the language power wisely.

**Comment by iAdramelk** [comment]
There is other usecases for using & without spacebars that @module didn't cover:
1. You may need to extend name for more than one class in selector: `&_type_active &__element` to `.b-block_type_active b-block__element` (for using modifiers in BEM).
2. `.bar&` used inside  `.foo`  must return `.bar.foo` (way to create multiclass and extend base class).
3. `button&` used inside `.foo` must exted it to `button.foo` (to make special case for buttons with this class).

From that I can see now it's looks like that using shorthand name is the simplest way to make all that usecases possible. (But I agree that it is ugly, and if there is a clearer way to do it I vill be happy).

**Comment by jamiebuilds** [comment]
I feel like the best way to implement this would to be just allow the & selector to represent the entire selector tree(?) so far.. Just allow the developer to put the & where ever they want. Then they can decide what system they want.

Otherwise it feels like complexity being added to Roole to just add limitations. This has long bothered me about sass/less and I like that Stylus does not do it this way.

I don't think Roole should try and correct the developer if they leave the space out.

**Comment by curvedmark** [comment]
Will this syntax satisfy your needs?

```
@module .menu with '-'
    position: absolute

    @module .item   
        display: block

        .content a
            padding: 10px

        .arrow::before
            content: '>'

        &.active
            background: gray
```

which compiles to

```
.menu {
    position: absolute;
}
    .menu-item {
        display: block;
    }
        .menu-item-content a {
            padding: 10px;
        }
        .menu-item-arrow::before {
            content: '>';
        }
        .menu-item.menu-item-active {
            background: gray;
        }
```

A rough rule is that `@module` specifies a simple class selector, optionally with a separator string, and all class selectors nested inside it will be prepend by that class selector joined with that separator string.

**Comment by jamiebuilds** [comment]
This is _way_ more complicated than need be.. No one's going to want to type `@module` a hundred times when they can just use `&` with anything else. People understand the idea and it's 100% flexible. Personally it's one of the best selling points to Stylus, and people complain all the time that SASS limits what you can do with `&`.

**Comment by curvedmark** [comment]
```
.menu
    position: absolute

    &-item   
        display: block

        &-content a
            padding: 10px

        &-arrow::before
            content: '>'

        &&-active
            background: gray
```

But without the first selector (imagine when the module has many rules that the first selector is very far from the currently selector), it's not even clear what kind of selectors you are specifying.

**Comment by jamiebuilds** [comment]
You wouldn't use the dash like that (&&-active), and it would be just as clear that you are working off of everything so far.

`.menu` then `&-item` becomes `.menu-item`
`.menu` then `&-item` then `&-content` becomes `.menu-item-content`
`.menu` then `&-item` then `&.active` becomes `.menu-item.active`

``` sass
.menu
    position: absolute

    &-item   
        display: block

        &-content a
            padding: 10px

        &-arrow::before
            content: '>'

        &.active
            background: gray
```

Then you could do complex things like:

``` sass
.menu
    & &-item
```

becomes

``` css
.menu .menu-item { }
```

or

``` sass
.foo, .bar, .hello, .world
    &:first-child, &:last-child
```

becomes

``` css
.foo:first-child, .foo:last-child, .bar:first-child, .bar:last-child, .hello:first-child, .hello:last-child, .world:first-child, .world:last-child { }
```

I have hundreds if not thousands of places where I use `&`'s to create complex selectors.

**Comment by curvedmark** [comment]
when you need to specify `.menu-item.menu-item-active`, aren't you supposed to write `&&-active` when the parent selector is `.menu-item`? although I admit `.menu-item-active` is probably enough and selectors like `&&-active` should be pretty rare.

I'm still not convinced to support this kind of `&` selector. when you see a selector like `&-content a`, and you need to know what it really is, you need to check its parent selectors. if one of them is a complex selector like `.menu .menu-item`, it could be really confusing (authors might not do that very often, but nothing prevents it).

with `@module` you could be rest assured that only a prefix name will be prepended to the class selector name, and a simple selector will not be turned into a complex one. to make it more predictable, i'm thinking something like this:

```
@module .menu
    .item.active .content
        background: gray
```

only `.item` and `.content` will be prefixed with `menu-`.

**Comment by jamiebuilds** [comment]
It is a weird use case if you are being that specific, you are much more likely to see something like below. But that's the whole point of ampersand (it doesn't judge), you can build any selector you want with it.

``` sass
.menu
    &-item
        // .menu-item
        &.active
            // .menu-item.active
        &:hover
            // .menu-item:hover
```

Or if you really wanted:

``` sass
.menu
    & &-item
        // .menu .menu-item
        &&-active
            // .menu .menu-item.menu-item-active 
        &:hover
            // .menu .menu-item:hover
```

All it takes to change this is three characters

**Comment by curvedmark** [comment]
I've decided to go with `@module` (#62), which also has the benefit of being able to namespace existing css frameworks.

**Comment by jamiebuilds** [comment]
Terrible decision, deal-breaker

**Comment by curvedmark** [comment]
@thejameskyle I understand your frustration with this decision, and I'm sorry if it seems too sudden.

I read you objections again:

> No one's going to want to type `@module` a hundred times when they can just use `&` with anything else.

You don't need to type `@module` a lot, only on class selectors that should also be the namespace, which actually saves you from having to type `&-` in all class selectors that need to be namespaced.

Also the reasons why you like `&` more:

> People understand the idea and it's 100% flexible.

I agree it's intuitive as to how it works, but its readability isn't very well.

> Personally it's one of the best selling points to Stylus, and people complain all the time that SASS limits what you can do with `&`.

Could you provide some links to these complains? Are they complaining it only because it made hard for them to do selector namespacing? Or Are there other use cases?

Also some technical reasons why I'm reluctant to support it:

Unlike Stylus, Roole (and also Sass if I'm not wrong) actually parses each selector (e.g., Roole knows `[type=button]` is the same as `[type = button]`, whereas Stylus treats them as different selectors, so  `@extends [type=button]` won't match `[type = button]`. Stylus simply does string replacement when resolving `&`, so it's very cheap, but to support it in Roole, every node representing different kinds of selectors needs to check for `&`, and when the parent selector is a compound selector (e.g., `.menu .menu-item`), the reconstruction of nodes could get really complex.

**Comment by curvedmark** [comment]
With dc10dcb36fc0584756dded2a8c0cb959d0295da0, this actually becomes cheap to implement if the parser is able to recognize this kind of `&` selector, but I'm not sure what the grammar should be. Is allowing `&` followed by an identifier at the start of each compound selector ([the definition of compound selector](http://dev.w3.org/csswg/selectors4/#structure)) enough? Or you guys have other creative usage for it?

**Comment by jamiebuilds** [comment]
What `&` definition is easier to explain and/or easier to implement?

**SASS:**
The selector above. Must come at the beginning of another selector or with a space before. The selector must valid at every level (regardless if there are properties at that level).

**Stylus:**
The selector above exactly as it is written.

Go ahead and decide for yourself

**Comment by curvedmark** [comment]
Landed in 5841d321c33dfd348fe79e2f7ac95006cb65f609
